### ISSUE TITLE
THREDDS 5.4 WMS is much slower than 4.6

### ISSUE THREAD

**Issue by billyz313** [Unidata/tds#406]
We are still running V4.6 for production because 5.4 is many times slower.  We want to move production to 5.4 but simply can't.  

We are using  Ubuntu 22.04.2 LTS, OpenJDK 64-Bit Server VM Temurin-11.0.18+10 (build 11.0.18+10, mixed mode), python 3.10

To reproduce I add a wms layer to a leaflet map for v4.6 https://csthredds.servirglobal.net/thredds/climateserv_aggregated.html?dataset=emodis-ndvi_eastafrica_250m_10dy 
and on a different leaflet map i add v5.4 https://threddsx.servirglobal.net/thredds/catalog/climateserv_aggregated.html?dataset=emodis-ndvi_eastafrica_250m_10dy

I also have them added as L.timeDimension.layer.wms because we animate the layers.  This is where it really makes a lot of difference.  Animation using THREDDS 5.4 currently seems not possible due to the lag of the responses.  

Any help and suggestions are welcome.

**Comment by mnlerman** [comment]
Hello, we had a user encounter a similar issue a few months ago and after some investigation, we found that they were able to improve the performance by removing "-ea" from the JAVA_OPTS in their tomcat setenv.sh. Can you test that to see if it helps improve the performance issues that you're seeing?

**Comment by billyz313** [comment]
@mnlerman Hi Megan,  thank you for the suggestion.  I just tried that and it didn't seem to make any difference.  Here are the options I have set, maybe there is something else that may need adjusting and i'm just missing it?

```
NORMAL="-Xmx16384m -Xms512m -server -XX:+UseParallelGC -Djava.awt.headless=true"
HEAP_DUMP="-XX:+HeapDumpOnOutOfMemoryError"
HEADLESS="-Djava.awt.headless=true"

JAVA_OPTS="$CONTENT_ROOT $NORMAL $HEAP_DUMP $HEADLESS $JAVA_PREFS_ROOTS"
```

I do realize I have headless set twice.  I was having an issue with thredds crashing when I would zoom the layer on the map and I saw adding headless might help.  I added it in the NORMAL variable and it stopped crashing.  Not long after that I realized it was actually already there in the HEADLESS variable.  Not sure how it stopped it from crashing or if it was just a coincidence.

**Comment by billyz313** [comment]
Still looking for suggestions.  We're concerned that support for V4.6 has ended and we can't get V5 to produce anywhere near the same speed.  At this point we're being forced to look into other options which I would prefer not to have to do.  Any help would be greatly appreciated.

**Comment by haileyajohnson** [comment]
@billyz313 have you tested performance with the latest snapshot release of the TDS?

**Comment by billyz313** [comment]
@haileyajohnson No, we are using 5.4, maybe I can convince them to give it a shot.  I should just be able to replace the .war file and all the configs and custom palettes will just be read in with no issue?

**Comment by billyz313** [comment]
@haileyajohnson I upgraded to V5.5 this morning to see if it helped.  Unfortunately there seems to be no improvement in performance.  It's a huge lag from V4.6 when using the wms endpoint which is what we use to animate the data on the map.

**Comment by tdrwenski** [comment]
Thanks for testing that out.

Can you send a bit more info to help us try to reproduce your issue? If possible can you give us:
- The catalog or catalog entry for the slow dataset
- The data file(s) for that dataset or a link where we can download them
- The exact WMS URL that is slow

**Comment by tdrwenski** [comment]
Thank you for sending the extra info! I see how slow it is on your server and I am seeing similar performance issues locally using your datasets.

It seems slow with other services like NCSS as well so I don't think it's only WMS with the issue. It looks like this issue is related to a performance issue we are working on fixing for version 5.5-- Enhancements (such as scale, offset, fill value) are not handled well in the current version and can cause performance issues for large datasets.

We will keep you updated on our progress on this issue!

**Comment by billyz313** [comment]
Thank you.  Just a quick question if you know.  Will disabling everything except wms help performance at all?  I'm guessing it wouldn't, but we only actively use the wms for ClimateSERV so I think turning the rest of the options off is prolly a good idea just in general.

**Comment by tdrwenski** [comment]
> Will disabling everything except wms help performance at all? I'm guessing it wouldn't, but we only actively use the wms for ClimateSERV so I think turning the rest of the options off is prolly a good idea just in general.

Unfortunately, I don't think that will help at all. The only thing I think would help is to not have enhancements (scale, offset, fill value) in your data, which probably isn't a feasible workaround. I hope we will have a fix for you soon!

**Comment by billyz313** [comment]
Is there a quick way I can turn off the enhancements to see what it does and if it is usable like that?

**Comment by tdrwenski** [comment]
I don't think there is any easy way to turn them off unless you change the data itself. For instance, with ncdump on one of your data files I see:
```
variables:
	double latitude(latitude) ;
		latitude:_FillValue = NaN ;
		latitude:long_name = "latitude" ;
		latitude:units = "degrees_north" ;
		latitude:axis = "Y" ;
	int time(time) ;
		time:long_name = "time" ;
		time:axis = "T" ;
		time:bounds = "time_bnds" ;
		time:units = "seconds since 1970-01-01T00:00:00+00:00" ;
		time:calendar = "proleptic_gregorian" ;
	double longitude(longitude) ;
		longitude:_FillValue = NaN ;
		longitude:long_name = "longitude" ;
		longitude:units = "degrees_east" ;
		longitude:axis = "X" ;
	byte ndvi(time, latitude, longitude) ;
		ndvi:_FillValue = 127b ;
		ndvi:long_name = "ndvi" ;
		ndvi:units = "unitless" ;
		ndvi:comment = "Maximum value composite over dekad defined by time_bnds" ;
		ndvi:add_offset = 0. ;
		ndvi:scale_factor = 0.01 ;
		ndvi:missing_value = 127b ;
	int time_bnds(time, nbnds) ;
		time_bnds:long_name = "time_bounds" ;
```
The attributes like `add_offset`, `scale_factor`, `missing_value`, and `_FillValue` are the enhancements I am referring to. If you are creating the netcdf files yourself you could test if things are faster without those attributes. Otherwise you can use tools like ncgen/ncdump to test removing them, but I guess that would not be an actual workaround but only useful for testing.

**Comment by billyz313** [comment]
Hi @tdrwenski , I was just checking back to see if y'all were able to publish a fix for this yet.  Also I did look into removing the enhancements, but it will not work in the rest of the system without these.

**Comment by tdrwenski** [comment]
Hi, we are still working on this performance issue. We will try to let you know if we have an update!

**Comment by billyz313** [comment]
Thank you!  I will see if we can get it deployed asap!

**Comment by billyz313** [comment]
@haileyajohnson We upgraded to 5.5 and unfortunately there is no change in the performance.

**Comment by ashokgj** [comment]
yes, we too are facing the same... 4.6 was a lot faster than 5.5

**Comment by tdrwenski** [comment]
We have a few more performance fixes that I think may help you. You can download the latest snapshot here: https://downloads.unidata.ucar.edu/tds/5.5/thredds-5.5-SNAPSHOT.war

Let us know if that helps or not!

**Comment by tdrwenski** [comment]
Hi @billyz313, have you had a chance to test performance with the latest snapshot yet?

**Comment by billyz313** [comment]
@tdrwenski yes, we installed the new 5.5 snapshot you mentioned and didn't seem to have any effect on the performance.  Some of the folks thought it got slower, but i think it's about the same.  So, we're still running 4.6 in production...

**Comment by haileyajohnson** [comment]
@billyz313 we used your data as a benchmark for our performance improvements, so it's surprising that it hasn't fixed your issues (not that we're doubting you, we can see that it's slow). In our own tests, performance on serving your data vis WMS was at least twice as fast with recent changes...
Is it possible that your network could be blocking/scanning/slowing something down? 
There are definitely things we could have overlooked here, but it be good to verify that it's not a network config issue.

**Comment by billyz313** [comment]
@haileyajohnson Do you think the virtual aggregation could be causing the delay?

**Comment by tdrwenski** [comment]
Very large joinExisting aggregations can be slow on the first request, as all files in the aggregation will be opened to get the coordinate info. We have an aggregation cache that persists this info, so it should be much faster on the second request. I don't believe I saw this behavior on your server, however, it seems consistently slow.

You may want to compare the performance of a single unaggregated file vs an aggregation if you want to be sure it's not the aggregation causing the slow down. If do you think your performance issues are related to the aggregation, there are a couple things you could try. The aggregation cache is scoured daily, but you can turn this off by setting the scour period to `-1 sec` (see [here](https://docs.unidata.ucar.edu/tds/5.5/userguide/tds_config_ref.html#aggregation-cache)). If each file in your joinExisting aggregation has one time value, you can also try using a [dateFormatMark ](https://docs.unidata.ucar.edu/netcdf-java/5.5/userguide/ncml_aggregation.html#extracting-date-coordinates-from-the-filename) to extract the value from the file name, so that the file won't need to be opened to get this info.

**Comment by billyz313** [comment]
@tdrwenski We're considering switching our files from netcdf to zarr.  The 5.x version supports zarr right?  And is the wms tiling a lot faster thru THREDDS using a zarr file or is it similar tiling speed as netcdf?

**Comment by haileyajohnson** [comment]
@billyz313 unfortunately Zarr is still very in beta in the TDS library, but you're welcome to try it out

**Comment by billyz313** [comment]
@haileyajohnson thank you, I think we should throw a few files in, cross our fingers, and see what it does :)

**Comment by tlvu** [comment]
Just curious, has the 5.5 release fixed this issue?

**Comment by billyz313** [comment]
@tlvu Unfortunately no.  

@tdrwenski We tried to deploy 5.6 snapshot and it failed.  Here are messages from the catalina.out, not sure if it will be helpful.  After it failed, he rolled back to 5.5,  Do you have any suggestions?

```
20-Nov-2024 15:23:43.819 INFO [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployWAR Deploying web application archive [/usr/local/apache-tomcat-8.5.85/webapps/thredds##5.6SS.war]
20-Nov-2024 15:23:44.504 WARNING [localhost-startStop-2] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:23:44.511 WARNING [localhost-startStop-2] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:23:47.074 INFO [localhost-startStop-2] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
20-Nov-2024 15:23:47.082 SEVERE [localhost-startStop-2] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
20-Nov-2024 15:23:47.083 SEVERE [localhost-startStop-2] org.apache.catalina.core.StandardContext.startInternal Context [/thredds##5.6SS] startup failed due to previous errors
20-Nov-2024 15:23:47.096 INFO [localhost-startStop-2] org.apache.catalina.startup.HostConfig.deployWAR Deployment of web application archive [/usr/local/apache-tomcat-8.5.85/webapps/thredds##5.6SS.war] has finished in [3,276] ms
20-Nov-2024 15:25:53.545 WARNING [http-nio-8080-exec-9] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:25:53.552 WARNING [http-nio-8080-exec-9] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:25:55.705 INFO [http-nio-8080-exec-9] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
20-Nov-2024 15:25:55.711 SEVERE [http-nio-8080-exec-9] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
20-Nov-2024 15:25:55.712 SEVERE [http-nio-8080-exec-9] org.apache.catalina.core.StandardContext.startInternal Context [/thredds##5.6SS] startup failed due to previous errors
20-Nov-2024 15:26:08.897 INFO [http-nio-8080-exec-10] org.apache.catalina.util.LifecycleBase.stop The stop() method was called on component [StandardEngine[Catalina].StandardHost[localhost].StandardContext[/thredds##5.6SS]] after stop() had already been called. The second call will be ignored.
20-Nov-2024 15:26:09.435 INFO [http-nio-8080-exec-10] org.apache.catalina.startup.HostConfig.undeploy Undeploying context [/thredds##5.6SS]
20-Nov-2024 15:30:47.162 INFO [localhost-startStop-3] org.apache.catalina.startup.HostConfig.deployWAR Deploying web application archive [/usr/local/apache-tomcat-8.5.85/webapps/thredds##56SS.war]
20-Nov-2024 15:30:47.808 WARNING [localhost-startStop-3] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:30:47.813 WARNING [localhost-startStop-3] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:30:49.952 INFO [localhost-startStop-3] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
20-Nov-2024 15:30:49.958 SEVERE [localhost-startStop-3] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
20-Nov-2024 15:30:49.959 SEVERE [localhost-startStop-3] org.apache.catalina.core.StandardContext.startInternal Context [/thredds##56SS] startup failed due to previous errors
20-Nov-2024 15:30:49.966 INFO [localhost-startStop-3] org.apache.catalina.startup.HostConfig.deployWAR Deployment of web application archive [/usr/local/apache-tomcat-8.5.85/webapps/thredds##56SS.war] has finished in [2,804] ms
20-Nov-2024 15:31:04.599 WARNING [http-nio-8080-exec-13] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:31:04.604 WARNING [http-nio-8080-exec-13] org.apache.tomcat.util.descriptor.web.WebXml.setVersion Unknown version string [5.0]. Default version will be used.
20-Nov-2024 15:31:06.708 INFO [http-nio-8080-exec-13] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
20-Nov-2024 15:31:06.713 SEVERE [http-nio-8080-exec-13] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
20-Nov-2024 15:31:06.714 SEVERE [http-nio-8080-exec-13] org.apache.catalina.core.StandardContext.startInternal Context [/thredds##56SS] startup failed due to previous errors
```

**Comment by lesserwhirls** [comment]
Greetings @billyz313! It looks like we upgraded from Spring 5 to Spring 6 in August, and that required a jump to Java 17 and Tomcat 10. Based on the path to tomcat in your log files, one or both of those might be the issue.

**Comment by billyz313** [comment]
@lesserwhirls Thank you, we are setting up a new machine that meets the requirements and will give it a shot on there.  We are hoping to see an increase in the WMS speed.
